File: 0136_single_number/single_number.cpp
#include <iostream>
#include <vector>

using namespace std;

/**
 * Time: O(n)
 * Space: O(1)
 */
int single_number(vector<int> &nums)
{
    int res = 0;
    for (auto i : nums)
        res ^= i;
    return res;
}

File: 0137_single_number_II/single_number.py
from collections import Counter
from typing import List


def single_number(nums: List[int]) -> int:
    """
    Time: O(n)
    Space: O(n)
    """
    c = Counter(x for x in nums)
    for k, v in c.items():
        if v == 1:
            return k


def single_number_bitwise(nums: List[int]) -> int:
    """
    Time: O(n)
    Space: O(1)
    """
    once = twice = 0
    for i in nums:
        once = (i ^ once) & ~twice
        twice = (i ^ twice) & ~once
    return once
